fibonacci.asm

```diff
@@ -65,24 +50,56 @@ head:	 .asciiz  "The Fibonacci numbers are:\n" 	# Print a little helpful intro
 
 	      .text
 	      
-print:	add  $t0, $zero, $a0  # starting address of array of data to be printed
-	      add  $t1, $zero, $a1  # initialize loop counter to array size
-	      la   $a0, head        # load address of the print heading string
-	      li   $v0, 4           # specify that you're printing a string
-	      syscall               # print the heading string
-
-out:	  lw   $a0, 0($t0)      # load the integer to be printed (the current Fib. number)
-	      li   $v0, 1           # specify that you're printing an integer
-	      syscall               # print fibonacci number
-
-	      la   $a0, space       # load address of spacer for syscall
-	      li   $v0, 4           # specify Print String service
-	      syscall               # print the spacer string
-
-	      addi $t0, $t0, 4      # increment address of data to be printed
-	      addi $t1, $t1, -1     # decrement loop counter
-	      bgtz $t1, out         # repeat while not finished
-
-	      jr   $ra              # return from subroutine
-	# End of subroutine to print the numbers on one line
-	###############################################################
+# Recursive fib(n) routine
+fib:      addi $sp, $sp, -8 # reserve space for two words on the stack
+ 	  sw $ra, 4($sp)   # save $ra on stack
+ 	  
+	  add $v0, $zero, $a0 # set the return value initially equal to n
+	  blt $a0, 2, fib_return # if n < 2, return
+	  
+	  sw $a0, 0($sp) # save n on the stack
+
+	  addi $a0, $a0, -1 # fib(n - 1)
+          jal fib
+          
+          lw $a0, 0($sp) # restore n
+          sw $v0, 0($sp) # save fib(n - 1) on the stack
+          
+          addi $a0, $a0, -2 # fib(n - 2)
+          jal fib
+          
+          lw $v1, 0($sp) # restore fib(n - 1)
+          
+          add $v0, $v0, $v1 # set the return value equal to fib(n - 1) + fib(n - 2)
+          
+fib_return: lw $ra, 4($sp) # restore the $ra
+	    addi $sp, $sp, 8 # restore the stack pointer
+            jr $ra # return from subroutine
+ 
+ 
+print:  addi $sp, $sp, -4 # reserve space for one word on the stack
+ 	sw $ra, 0($sp)   # save $ra on stack
+ 	  
+        add $t0, $zero, $a0 # get number to print from argument
+        li $t1, 1  # initialize loop counter to 1
+	la $a0, head # load address of the print heading string
+	print_str
+	
+out:    add $a0, $zero, $t1 # pass the loop counter to fib
+        jal fib
+        
+        add $a0, $zero, $v0 # load the integer to be printed (the current Fib. number)
+	print_int
+	
+	la $a0, space # load address of spacer for syscall
+	print_str
+	
+	addi $t1, $t1, 1 # increment the loop counter
+	ble $t1, $t0, out # repeat while not finished
+	
+	lw $ra, 0($sp) # restore the $ra
+	addi $sp, $sp, 4 # restore the stack pointer
+	jr $ra # return from subroutine
+	
+# End of subroutine to print the numbers on one line
+###############################################################
```

Comment:
You demonstrate considerable skill in MIPS and I can appreciate your recursive implementation. however I must comment that the recursive design is MUCH less efficient than the dynamic programming approach.